### PR TITLE
ISSUE #870: ScanAndCompareGarbageCollector: harden against LM bugs

### PR DESCRIPTION
**Pull request by revans2**
This is a port of https://github.com/apache/bookkeeper/issues/870

### CODE REVIEW

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/GcLedgersTest.java [hunk 1]
```diff
@@ -300,4 +315,328 @@ public void clean(long ledgerId) {
...
+
+        garbageCollector.gc(cleaner);
+        assertEquals("Should have cleaned all the created ledgers", createdLedgers, cleaned);
+    }
+
+    /*
+     * in this test scenario no created ledger is deleted, but ledgeriterator is screwed up and returns hasNext to be
+     * false and next to be null. So even in this case it is expected not to clean any ledger's data.
+     *
+     * This testcase is needed for validating fix of bug - W-4292747.
```
Comment (agresch) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/GcLedgersTest.java]: is this a valid tracking issue?  How does one find it if so?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/GcLedgersTest.java [hunk 2]
```diff
@@ -300,4 +315,328 @@ public void clean(long ledgerId) {
...
+            public void clean(long ledgerId) {
+                LOG.info("Cleaned {}", ledgerId);
+                cleaned.add(ledgerId);
+            }
+        };
+
+        validateLedgerRangeIterator(createdLedgers);
+
+        garbageCollector.gc(cleaner);
+        assertTrue("Should have cleaned nothing", cleaned.isEmpty());
```
Comment (agresch) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/GcLedgersTest.java]: are there any tests with baseConf.setVerifyMetadataOnGc(true) that actually do any cleaning, just to validate it doesn't disable cleaning altogether?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/GcLedgersTest.java [hunk 3]
```diff
@@ -300,4 +315,328 @@ public void clean(long ledgerId) {
...
+    }
+
+    /*
+     * in this test scenario no created ledger is deleted, but ledgeriterator is screwed up and returns hasNext to be
+     * false and next to be null. So even in this case it is expected not to clean any ledger's data.
+     *
+     * This testcase is needed for validating fix of bug - W-4292747.
+     *
+     * ScanAndCompareGarbageCollector/GC should clean data of ledger only if both the LedgerManager.getLedgerRanges says
+     * that ledger is not existing and also ledgerManager.readLedgerMetadata fails with error
```
Comment (agresch) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/GcLedgersTest.java]: possibly add this test with baseConf.setVerifyMetadataOnGc(false) to validate existing functionality remains the same?

**Review by agresch — APPROVED** [review]
Possible further tests for completeness, but I'm fine with it as is.

**Review by kishorvpatil — APPROVED** [review]
👍

**Review by rdhabalia — APPROVED** [review]
👍

**Pull request merged**